<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Цель-исследования" data-toc-modified-id="Цель-исследования-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Цель исследования</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Структура-базы-данных" data-toc-modified-id="Структура-базы-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Структура базы данных</a></span></li><li><span><a href="#Задачи-исследования" data-toc-modified-id="Задачи-исследования-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Задачи исследования</a></span></li><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Подключение-к-базе-данных" data-toc-modified-id="Подключение-к-базе-данных-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a href="#Выполнение-задач" data-toc-modified-id="Выполнение-задач-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Выполнение задач</a></span><ul class="toc-item"><li><span><a href="#Задача-№-1" data-toc-modified-id="Задача-№-1-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>Задача № 1</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-8.1.1"><span class="toc-item-num">8.1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Задача-№-2" data-toc-modified-id="Задача-№-2-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>Задача № 2</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-8.2.1"><span class="toc-item-num">8.2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Задача-№-3" data-toc-modified-id="Задача-№-3-8.3"><span class="toc-item-num">8.3&nbsp;&nbsp;</span>Задача № 3</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-8.3.1"><span class="toc-item-num">8.3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Задача-№-4" data-toc-modified-id="Задача-№-4-8.4"><span class="toc-item-num">8.4&nbsp;&nbsp;</span>Задача № 4</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-8.4.1"><span class="toc-item-num">8.4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Задача-№-5" data-toc-modified-id="Задача-№-5-8.5"><span class="toc-item-num">8.5&nbsp;&nbsp;</span>Задача № 5</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-8.5.1"><span class="toc-item-num">8.5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# SQL

## Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Цель исследования

Проанализировать базу данных и на основе анализа сформулировать ценностное предложение для нового продукта

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Структура базы данных

![Структура базы данных](https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?table=block&id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=2000&userId=&cache=v2)

## Задачи исследования

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так будут исключены из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Импорт библиотек

In [1]:
import pandas as pd
from sqlalchemy import create_engine

## Подключение к базе данных

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
# Формирует строку для подключения
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db']) 

Чтобы создать соединение, строку с данными для подключения передаём функции **create_engine()**

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

Готово! К базе подключились, теперь можно извлекать результаты SQL-запроса

## Выполнение задач

Для начала исследуем таблицы, извлекём по 5 строк из каждой. Запрос запишем в переменную **query** и через цикл выведем информацию из каждой таблицы

In [4]:
for table_name in  ['books','ratings', 'reviews', 'publishers', 'authors']:
    query = '''SELECT *
               FROM ''' + table_name 
    print('Таблица ', table_name)
    display(pd.read_sql_query(query, con = engine))
    print(pd.read_sql_query(query, con = engine).info())

Таблица  books


book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
Таблица  ratings


rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
Таблица  reviews


review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
Таблица  publishers


publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
Таблица  authors


author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None


### Задача № 1

Посчитать, сколько книг вышло после 1 января 2000 года

In [5]:
query = '''
SELECT COUNT(book_id)
FROM books
WHERE CAST(publication_date AS DATE) >= '2000-01-01';
'''
display(pd.read_sql_query(query, con = engine))

count
0    821

#### Вывод

В данном вопросе использовали простую струтуру с условным оператором WHERE, получили 821 книга вышла после 1 января 2000 года

### Задача № 2

Для каждой книги посчитать количество обзоров и среднюю оценку

In [6]:
query = '''
SELECT b.title,
       COUNT(DISTINCT(rv.review_id)) AS count_review,
       ROUND(AVG(rt.rating),2) AS avg_rating
FROM books AS b
LEFT OUTER JOIN reviews AS rv ON b.book_id = rv.book_id
LEFT OUTER JOIN ratings AS rt ON b.book_id = rt.book_id
GROUP BY b.book_id;
'''
display(pd.read_sql_query(query, con = engine))
print(pd.read_sql_query(query, con = engine).info())

title  count_review  \
0                                         'Salem's Lot             2   
1                   1 000 Places to See Before You Die             1   
2    13 Little Blue Envelopes (Little Blue Envelope...             3   
3    1491: New Revelations of the Americas Before C...             2   
4                                                 1776             4   
..                                                 ...           ...   
995           Wyrd Sisters (Discworld  #6; Witches #2)             3   
996                        Xenocide (Ender's Saga  #3)             3   
997                                    Year of Wonders             4   
998                        You Suck (A Love Story  #2)             2   
999  Zen and the Art of Motorcycle Maintenance: An ...             4   

     avg_rating  
0          3.67  
1          2.50  
2          4.67  
3          4.50  
4          4.00  
..          ...  
995        3.67  
996        3.40  
997        3.20  
998        4.50  
999        3.83  

[1000 rows x 3 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   title         1000 non-null   object 
 1   count_review  1000 non-null   int64  
 2   avg_rating    1000 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 23.6+ KB
None


#### Вывод

В данном запросе использовали объединение таблиц с помощью оператора LEFT OUTER JOIN, что бы не потерять книги при объединении, так как не на все книги написаны отзывы и приведены оценки.

### Задача № 3

Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так будут исключены из анализа брошюры

In [7]:
query = '''
SELECT publisher, COUNT(publisher)
FROM books AS b
JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE num_pages > 50
GROUP BY publisher
ORDER BY COUNT(publisher) DESC
LIMIT 1;
'''
display(pd.read_sql_query(query, con = engine))

publisher  count
0  Penguin Books     42

#### Вывод

Здесь в запросе использовали объединение, комбинированное с условным оператором и группировкой, получили, что издательство Penguin Books выпустило наибольшее число книг(42) штуки, больше 59 листов.

### Задача № 4

Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [8]:
query = '''
SELECT a.author, AVG(r.rating)
FROM books AS b
JOIN authors AS a ON b.author_id = a.author_id
JOIN ratings AS r ON b.book_id = r.book_id
GROUP BY author
HAVING COUNT(r.rating_id) > 50 
ORDER BY AVG(r.rating) DESC
LIMIT 1;
'''
display(pd.read_sql_query(query, con = engine))

author       avg
0  J.K. Rowling/Mary GrandPré  4.288462

#### Вывод

Здесть всё тоже самое, за исключением добавления условного оператора после группировки по авторам. J.K. Rowling/Mary GrandPré автор с самой высокой средней оценкой книг.

### Задача № 5

Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [9]:
query = '''
SELECT ROUND(AVG(ctr.count_rv), 2)
FROM (SELECT rt.username, 
             COUNT(DISTINCT(rt.rating_id)) AS count_rt, 
             COUNT(DISTINCT(rv.review_id)) AS count_rv 
      FROM ratings AS rt
      FULL OUTER JOIN reviews AS rv ON rv.username = rt.username
      GROUP BY rt.username
      HAVING COUNT(DISTINCT(rt.rating_id)) > 50) AS ctr;
'''
display(pd.read_sql_query(query, con = engine))

round
0  24.33

#### Вывод

В даннном запросе пришлось использовать вложенный запрос, хотя можно было использовать и оконную функцию. В итоге получилось, что пользователи, которые поставили больше 50 оценок, в среднем оставляли по 24 отзыва.

## Общий вывод

В процессе выполнения проекта мы подключились к внешней базе данный с помощью Phyton, написали, разного уровня сложности, SQL запросы, что даёт понять, что с базами данных легко работать с помощью Phyton и это достаточно мощный инструмент, который может интегрироваться в разные среды, для работы с данными и не только.

Все 5, поставленных задач, были решены,мы узнали, что :
- после 2000 года выпущена большая часть книг из предоставленных данных;
- издательство **Penguin Books** выпустило больше всех книг, а именно 42 штуки;
- автор **J.K. Rowling/Mary GrandPré** выставляет самые большие оценки книгам, в среднем это оценка 4.288;
- 24 обзора - именно столько в среднем обзоров у пользователей, которые оставили больше 50 оценок книгам.